### PR TITLE
Move protocolsettings checks to the validating webhook

### PR DESCRIPTION
**Pull request by raypinto**
**Description**

Changes proposed in this pull request:

- Move QoS, webhook Auth type and grant type validations to the validating webhook 

**Related issue(s)**

### CODE REVIEW

File: components/eventing-controller/pkg/ems/api/events/types/grant_type.go [hunk 1]
```diff
@@ -3,3 +3,11 @@ package types
...
 
 const GrantTypeClientCredentials GrantType = "client_credentials"
+
+func IsInvalidGrantType(value string) bool {
+	return value != string(GrantTypeClientCredentials)
+}
+
+func GetGrantType(_ string) GrantType {
+	return GrantTypeClientCredentials
+}
```
Comment (marcobebway) [inline components/eventing-controller/pkg/ems/api/events/types/grant_type.go:13]: Why do we need this function? We can use the already exported const `GrantTypeClientCredentials`.

File: components/eventing-controller/pkg/ems/api/events/types/auth_type.go [hunk 2]
```diff
@@ -5,3 +5,11 @@ type AuthType string
...
 	AuthTypeClientCredentials AuthType = "oauth2"
 )
+
+func IsInvalidAuthType(value string) bool {
+	return value != string(AuthTypeClientCredentials)
+}
+
+func GetAuthType(_ string) AuthType {
+	return AuthTypeClientCredentials
+}
```
Comment (marcobebway) [inline components/eventing-controller/pkg/ems/api/events/types/auth_type.go:15]: Why do we need this function? We can use the already exported const `AuthTypeClientCredentials`.

File: components/eventing-controller/api/v1alpha2/subscription_webhook.go [hunk 3]
```diff
@@ -155,6 +170,13 @@ func (s *Subscription) validateSubscriptionSink() *field.Error {
 	return nil
 }
 
+func (s *Subscription) ifKeyExistsInConfig(key string) bool {
+	if _, ok := s.Spec.Config[key]; ok {
+		return true
+	}
+	return false
```
Comment (marcobebway) [inline components/eventing-controller/api/v1alpha2/subscription_webhook.go:177]: No need for the if-statement, the following code is enough:

```golang
func (s *Subscription) ifKeyExistsInConfig(key string) bool {
    _, ok := s.Spec.Config[key]
    return ok
}
```

File: components/eventing-controller/api/v1alpha2/subscription_webhook.go [hunk 4]
```diff
@@ -115,11 +117,24 @@ func (s *Subscription) validateSubscriptionTypes() *field.Error {
...
+	if s.ifKeyExistsInConfig(WebhookAuthType) && types.IsInvalidAuthType(s.Spec.Config[WebhookAuthType]) {
+		allErrs = append(allErrs, MakeInvalidFieldError(ConfigPath, s.Name, InvalidAuthTypeErrDetail))
+	}
+	if s.ifKeyExistsInConfig(WebhookAuthGrantType) && types.IsInvalidGrantType(s.Spec.Config[WebhookAuthGrantType]) {
+		allErrs = append(allErrs, MakeInvalidFieldError(ConfigPath, s.Name, InvalidGrantTypeErrDetail))
+	}
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return allErrs
```
Comment (marcobebway) [inline components/eventing-controller/api/v1alpha2/subscription_webhook.go:137]: I think we can remove the if-statement so just always return `return allErrs`.

File: components/eventing-controller/pkg/ems/api/events/types/grant_type.go (hunk 1 above)
Comment (raypinto) [reply components/eventing-controller/pkg/ems/api/events/types/grant_type.go:13]: The idea was to keep it extensible for future. If there are more values tomorrow, we pass the string value as a parameter and then return the `GrantType` value

File: components/eventing-controller/pkg/ems/api/events/types/auth_type.go (hunk 2 above)
Comment (raypinto) [reply components/eventing-controller/pkg/ems/api/events/types/auth_type.go:15]: The idea was to keep it extensible for future. If there are more values tomorrow, we pass the string value as a parameter and then return the `AuthType` value

File: components/eventing-controller/api/v1alpha2/subscription_webhook.go (hunk 3 above)
Comment (raypinto) [reply components/eventing-controller/api/v1alpha2/subscription_webhook.go:177]: that's right! will change

File: components/eventing-controller/api/v1alpha2/subscription_webhook.go (hunk 4 above)
Comment (raypinto) [reply components/eventing-controller/api/v1alpha2/subscription_webhook.go:137]: I think that's right! An empty slice would be nil (need to double check it)

Comment (raypinto) [reply components/eventing-controller/api/v1alpha2/subscription_webhook.go:137]: done

File: components/eventing-controller/api/v1alpha2/subscription_webhook.go (hunk 3 above)
Comment (raypinto) [reply components/eventing-controller/api/v1alpha2/subscription_webhook.go:177]: done

**Review by marcobebway — APPROVED** [review]

**Review by marcobebway — APPROVED** [review]

**Pull request merged**